# FINM 32000 Numerical Methods HW3

---- Yitong Wang

In [2]:
import numpy as np

## Problem 1

- Consider a portfolio long one $(K,T_2)$-forward contract and short one $(F_t,T2)$-forward contract, note that:
    - $(F_t,T2)$-forward contract has time-t value 0 ('fair price', settled exactly at time $t$)

    - $(K,T_2)$-forward contract has time-t value $f_t$ ($K$ can be settled in the past)

    - at time $t$, the portfolio has value $1 \cdot f_t - 1 \cdot 0 = f_t$ 

    - at time $T_2$, the portfolio pays $1 \cdot (S_{T_2} - K) - 1 \cdot (S_{T_2} - F_t) = F_t - K$

    - thus, $f_t = (F_t - K) \cdot e^{-r(T_2-t)}$

- $f_t = (F_t - K) \cdot e^{-r(T_2-t)}$ is the time-t value of the $(K,T_2)$-forward contract (which is written before t, while the contract written at time has delivery price exactly to be $F_t$), note that with $t$ changing, $F_t$ changes. Thus, once written, the value of a forward contract changes to market.

- We can derive $F_t$ if the underlying dynamics is specified:
    - the time-t value of $(F_t,T2)$-forward contract should be $e^{-r(T_2-t)}\mathbb{E}_t(S_{T_2}-F_t)$, and should be 0 (where $\textcolor{red}{\mathbb{E}}$ denotes for $\textcolor{red}{\text{risk-neutral}}$)

    - $e^{-r(T_2-t)}\mathbb{E}_t(S_{T_2}-F_t)=0 \Rightarrow F_t = \mathbb{E}_t(S_{T_2})$ 

    - note that $F_t$ is a martingale with final level $F_{T_2} = S_{T_2}$ (at the delivery date $T_2$, the forward price must equal to the underlying asset price)
    
    - if underlying $S$ is a non-dividend stock: $F_t = S_t \cdot e^{r(T_2-t)}$. This can be derived using non-abritrage pricing. If, say, $F_t> S_t \cdot e^{r(T_2-t)}$, then arbitrage would exist:
        - At time $t$, borrow $S_t$ dollars, buy the stock, and short the forward (with delivery price $F_t$ and time-t value 0)
        - At time $T_2$, deliver the stock, and receive $F_t$, which is more than enough to cover your accumulated debt of $S_t \cdot e^{r(T_2-t)}$ dollars
    
    - However, if $S$ is the spot price of a barrel of crude oil (so, for all $t$, the time-t price for time-t delivery is $S_t$ per barrel), then this argument fails. This is because in this situation, we can not use the above statement of "At time $T_2$, deliver the curde oil, and receive $F_t$, which is ~~more than enough~~ to cover your accumulated debt of $S_t \cdot e^{r(T_2-t)}$ dollars". Since now the accumulated debt is no longer $S_t \cdot e^{r(T_2-t)}$ dollars due to the <font color='skyblue'> storage costs </font> incurred by curde oil.

    - Thus, for crude oil, we need more assumptions to relate $F_t$ and $S_t$. The $S$ denotes spot crude oil, and $F_t$ denotes the time-t forward price for T2-delivery crude oil). One approach is to model the risk-neutral dynamics of $S$. Assume that $S$ satisfies (Where $W$ is BM under risk-neutral measure):
    $$ S_t = exp(X_t) $$
    $$ dX_t = \kappa (\alpha -X_t) dt + \sigma dW_t $$
    
    $$ \Rightarrow F_t= \mathbb{E}_t(S_{T_2}) =  \exp \left[ e^{-\kappa (T_2 - t)} \log S_t + (1 - e^{-\kappa (T_2 - t)}) \alpha + \frac{\sigma^2}{4 \kappa} (1 - e^{-2 \kappa (T_2 - t)}) \right] $$
    



#### (a) 
$f_t = (F_t - K) \cdot e^{-r(T_2-t)}$
#### (b)
we can not use the above statement of "At time $T_2$, deliver the curde oil, and receive $F_t$, which is ~~more than enough~~ to cover your accumulated debt of $S_t \cdot e^{r(T_2-t)}$ dollars". Since now the accumulated debt is no longer $S_t \cdot e^{r(T_2-t)}$ dollars due to the <font color='skyblue'> storage costs and transaction costs</font> incurred by curde oil.

#### (c) & (d)
- Suppose $\kappa= 0.472, \alpha= 4.4, \sigma= 0.368, r= 0.05, S_0= 106.9$.

- Let $C$ be the  time-0 price of a $K$-strike $T_1$-expiry European call on $F$.   

- So this  call  pays $(F_{T_1}−K)^+$. Let the call option have strike $K= 103.2$ and expiration $T_1= 0.5$. Let the forward have delivery date $T_2= 0.75$.  

$$ \because F_t= \mathbb{E}_t(S_{T_2}) =  \exp \left[ e^{-\kappa (T_2 - t)} \log S_t + (1 - e^{-\kappa (T_2 - t)}) \alpha + \frac{\sigma^2}{4 \kappa} (1 - e^{-2 \kappa (T_2 - t)}) \right] $$

$$ \therefore F_{T_1} = \mathbb{E}_{T_1}(S_{T_2}) =  \exp \left[ e^{-\kappa (T_2 - T_1)} \log S_{T_1} + (1 - e^{-\kappa (T_2 - T_1)}) \alpha + \frac{\sigma^2}{4 \kappa} (1 - e^{-2 \kappa (T_2 - T_1)}) \right] $$

- Thus need to simulate $S_t$ to derive $S_{T_1}$ to derive $F_{T_1}$ to derive $(F_{T_1}−K)^+$

In [40]:
# Exponential Ornstein-Uhlenbeck process

class XOU:

    def __init__(self, kappa, alpha, sigma, S0, r):

        self.kappa = kappa
        self.alpha = alpha
        self.sigma = sigma
        self.S0 = S0
        self.r = r
    

In [41]:
class CallOnForwardPrice:

    def __init__(self, K1, T1, T2):

        self.K1 = K1
        self.T1 = T1
        self.T2 = T2


In [42]:
class MCengine:

    def __init__(self, N, M, epsilon, seed):

        self.N = N   # Number of timesteps on each path
        self.M = M   # Number of paths
        self.epsilon = epsilon  # For the dC/dS calculation
        self.rng = np.random.default_rng(seed=seed) # Seeding the random number generator with a specified number helps make the calculations reproducible

    def price_call_XOU(self, contract, dynamics):

        alpha = dynamics.alpha
        kappa = dynamics.kappa
        sigma = dynamics.sigma
        S0 = dynamics.S0
        r = dynamics.r

        K1 = contract.K1
        T2 = contract.T2
        T1 = contract.T1
        dt = T1/self.N

        S_lst = []
        S_up_lst = []
        for m in range(self.M): # generate S_{T1} for M times
            Xt = np.log(S0) # S0 = exp(X0) thus X0 = ln(S0)
            Xt_up = np.log(S0+self.epsilon)
            for n in range(self.N): # N times to derive for S_{T1}
                random_norm = self.rng.normal() # self.rng.normal() generates pseudo-random normals
                dX = kappa*(alpha-Xt)*dt + sigma*random_norm*np.sqrt(dt) # don't forget np.sqrt(dt)!
                Xt += dX
                dX_up = kappa*(alpha-Xt_up)*dt + sigma*random_norm*np.sqrt(dt)
                Xt_up += dX_up
            S = np.exp(Xt)
            S_up = np.exp(Xt_up)
            S_lst.append(S)
            S_up_lst.append(S_up)
        
        def StoF(S, kappa, alpha, sigma, T1, T2):
            F = np.exp( np.exp(-kappa*(T2-T1)) * np.log(S) + (1 - np.exp(-kappa*(T2-T1)))*alpha + sigma**2/(4*kappa)*(1-np.exp(-2*kappa*(T2-T1))) )
            return F

        payoff_lst = [max(StoF(S, kappa, alpha, sigma, T1, T2)-K1, 0) for S in S_lst] # list size = M
        payoff_up_lst = [max(StoF(S_up, kappa, alpha, sigma, T1, T2)-K1, 0) for S_up in S_up_lst]
        
        payoff_array = np.array(payoff_lst)
        payoff_up_array = np.array(payoff_up_lst)

        call_price = np.exp(-r*(T1)) * np.mean(payoff_array)
        call_price_up = np.exp(-r*(T1)) * np.mean(payoff_up_array)
        
        std_sample = np.std(payoff_array, ddof=1)
        standard_error = std_sample/np.sqrt(self.M)
        
        call_delta = (call_price_up-call_price)/self.epsilon

        return(call_price, standard_error, call_delta)


In [37]:
hw5dynamics=XOU(kappa = 0.472, alpha = 4.4, sigma = 0.368, S0 = 106.9, r = 0.05)

hw5contract=CallOnForwardPrice(K1 = 103.2, T1 = 0.5, T2 = 0.75)

hw5MC = MCengine(N=100, M=100000, epsilon=0.01, seed=0)
# Change M if necessary

(call_price, standard_error, call_delta) = hw5MC.price_call_XOU(hw5contract,hw5dynamics)

print(f"call price:{round(call_price, 4)}; standard error: {round(standard_error, 4)}; call delta: {round(call_delta, 4)}")

call price:7.7323; standard error: 0.0432; call delta: 0.3396


#### (e)
$$\because f_t = (F_t - K) \cdot e^{-r(T_2-t)}$$

$$\therefore \frac{df_t}{dS} = \frac{dF_t}{dS} \cdot e^{-r(T_2-t)}$$

$$\because F_t= \mathbb{E}_t(S_{T_2}) =  \exp \left[ e^{-\kappa (T_2 - t)} \log S_t + (1 - e^{-\kappa (T_2 - t)}) \alpha + \frac{\sigma^2}{4 \kappa} (1 - e^{-2 \kappa (T_2 - t)}) \right] $$

$$\therefore \frac{dF_t}{dS} =  F_t \frac{e^{-\kappa (T_2 - t)}}{S_t} $$

$$\therefore \frac{df_t}{dS} = \frac{dF_t}{dS} \cdot e^{-r(T_2-t)} =  \frac{F_t}{S_t} \cdot e^{-(r+\kappa)(T_2-t)}$$

$$\therefore \frac{df_0}{dS} = \frac{df_t}{dS}\bigg|_{t=0} =  \frac{F_t}{S_t} \cdot e^{-(r+\kappa)(T_2-t)} \bigg|_{t=0} = \frac{F_0}{S_0} \cdot e^{-(r+\kappa)T_2}$$

- plug $F_0$ into the above formula:

$$ \because F_t= \mathbb{E}_t(S_{T_2}) =  \exp \left[ e^{-\kappa (T_2 - t)} \log S_t + (1 - e^{-\kappa (T_2 - t)}) \alpha + \frac{\sigma^2}{4 \kappa} (1 - e^{-2 \kappa (T_2 - t)}) \right] $$

$$ \therefore F_0= \exp \left[ e^{-\kappa T_2} \log S_0 + (1 - e^{-\kappa T_2 }) \alpha + \frac{\sigma^2}{4 \kappa} (1 - e^{-2 \kappa T_2}) \right] $$



In [43]:
def calc_forward_delta_XOU(dynamics, contract):
    alpha, kappa, sigma, S0, r  = dynamics.alpha, dynamics.kappa, dynamics.sigma, dynamics.S0, dynamics.r

    T2 = contract.T2

    F0 = np.exp( np.exp(-kappa*T2)*np.log(S0) + (1- np.exp(-kappa*T2))*alpha + sigma**2/(4*kappa)*(1-np.exp(-2*kappa*T2)))
    forward_delta = F0/S0 * np.exp(-(r+kappa)*T2)

    return forward_delta
    

In [44]:
print(f"forward delta at t0: {round(calc_forward_delta_XOU(hw5dynamics, hw5contract),4)}")

forward delta at t0: 0.6465


#### <span style = "background-color: yellow">(f)</span>
- Suppose you want to hedge a position short one call (so your hedge portfolio should replicate aposition long one call), by continuously rebalancing a position in $T_2$-delivery forward contracts. 

- Thus, your hedge portfolio at time 0 should be long the call's time-0 delta ($C$ to $f$) shares of forward contracts.

$$ \frac{\partial{C}}{\partial{f}} = \frac{\partial{C}}{\partial{S}} \div \frac{df}{dS} $$


In [32]:
def calc_calltoforward_delta_XOU(call_delta, dynamics, contract):
    return call_delta / calc_forward_delta_XOU(dynamics, contract)

In [33]:
print(f'shares of contract to hedging: {round(calc_calltoforward_delta_XOU(call_delta, hw5dynamics, hw5contract),4)}')

shares of contract to hedging: 0.5253


#### <span style = "background-color: yellow">(g)</span>
The time-$T_2$ value of the portfolio is:
$$
V_{T_2} = \theta (F_{T_2} - K) + (\theta_{\text{max}} - \theta) (S_{T_1} - K)\\
 = \theta f_{T_2} + (\theta_{\text{max}} - \theta) C_{T_1}
$$

So the time-0 value of our contract is the expectation of the replication at time-0.
$$
V_0 = E_0[V_{T_2}] = \theta f_0 + (\theta_{\text{max}} - \theta) C_0 e^{-r(T_2-T_1)}\\
f_0 = (F_0 - K) e^{-rT_2}
$$

Therefore, we long 4,000 forward contracts and long 1,000 call options.

In [47]:
F_0 = np.exp(np.exp(-hw5dynamics.kappa*(hw5contract.T2))*np.log(hw5dynamics.S0)+hw5dynamics.alpha*(1-np.exp(-hw5dynamics.kappa*(hw5contract.T2)))+hw5dynamics.sigma**2/(4*hw5dynamics.kappa)*(1-np.exp(-2*hw5dynamics.kappa*(hw5contract.T2))))
f_0 = np.exp(-hw5dynamics.r*(hw5contract.T2)) * (F_0-hw5contract.K1)

print('purchase_agreement_contract value: ', 4000*f_0+1000*call_price)


purchase_agreement_contract value:  3996.4233731151935


### Problem 2

#### (a)
$$ dS_t = rS_tdt + \sigma (t) S_t dW_t $$

- The dynamics is capable of generating a non-constant (with respect to T) term-structure of implied volatility, since $\sigma (t)$ is changing when $t$ changes.
- The dynamics is not capable of generating an implied volatility skew (non-constant with respect to $K$), since $\sigma (t)$ is not changing when $S_t$ changes.

- Specifically, if $\sigma$ is a non-random function of t and $$ dS_t = rS_tdt + \sigma (t) S_t dW_t $$ then: $$ d logS_t = (r-\frac{1}{2}\sigma^2(t))dt + \sigma(t) dWt $$
so: $$ log S_t = log S_0 + \int_{0}^{T} (r-\frac{1}{2}\sigma^2(t))dt + \int_{0}^{T} \sigma(t) dW_t $$
where $ \int_{0}^{T} (r-\frac{1}{2}\sigma^2(t))dt$ is a normal integral and $ \int_{0}^{T} \sigma(t) dW_t$ is a Ito integral with the function inside to be a non-random function

- remember that for that kind of Ito integral, we have: $$\int_{0}^{T} \sigma(t) dW_t \sim N(0, \int_{0}^{T} \sigma^2(t) dt)$$

- Thus:
$$ log S_t = log S_0 + \int_{0}^{T} (r-\frac{1}{2}\sigma^2(t))dt \int_{0}^{T} + \sigma(t) dW_t = logS_0 + (r- \frac{\bar{ \sigma }_T^2}{2})T + \int_{0}^{T} \sigma(T) dW_t \\ \sim N(logS_0 + (r- \frac{\bar{\sigma}_T^2}{2})T, \bar{\sigma}_T^2 T) $$ 
- where: $$\bar{\sigma}_T = \sqrt{\frac{1}{T} \int_{0}^{T} \sigma^2(t) dt} $$

- Thus: $\sigma_{imp}(K,T) = \bar{\sigma}_T$

- Note that in this case $\sigma_{imp}(K,T) = \bar{\sigma}_T $ will change when $T$ changes, but will not change when $K$ changes. 

#### (b)

$$\sigma_{imp}(K,T) = \bar{\sigma}_T = \sqrt{\frac{1}{T} \int_{0}^{T} \sigma^2(t) dt}$$

$$\Rightarrow \sigma_{imp}^2 \cdot T =  \int_{0}^{T} \sigma^2(t) dt$$


- Thus, A time-varying local volatility function can be constructed as:
$$
\sigma^2(t) = \begin{cases} 
\sigma_{1, imp}^2 & \text{if } 0 \leq t \leq t_1 \\
\frac{\sigma_{2, imp}^2 t_2 - \sigma_{1, imp}^2 t_1}{t_2-t_1} & \text{if } t_1 < t \leq t_2 \\
\frac{\sigma_{3, imp}^2 t_3 - \sigma_{2, imp}^2 t_2}{t_3-t_2} & \text{if } t_2 < t \leq t_3
\end{cases}
$$

In [54]:
import numpy as np
from scipy.stats import norm
from scipy.optimize import bisect, brentq
from copy import copy

In [49]:
class CallOption:

    def __init__(self, K, T, price=None):
        self.K = K
        self.T = T
        self.price = price

In [51]:
class GBMdynamics:

    def __init__(self, S, r, rGrow, sigma=None):
        self.S = S # S0
        self.r = r # interest rate
        self.rGrow = rGrow # R_grow
        self.sigma = sigma # instantenous vol

    def update_sigma(self, sigma):
        self.sigma = sigma
        return self

In [48]:
class AnalyticEngine:

    def __init__(self):
        pass

    def BSpriceCall(self, dynamics, contract):
        # ignores contract.price if given, because this function calculates price based on the dynamics

        F = dynamics.S*np.exp(dynamics.rGrow*contract.T)
        std = dynamics.sigma*np.sqrt(contract.T)
        d1 = np.log(F/contract.K)/std+std/2
        d2 = d1-std
        call_price = np.exp(-dynamics.r*contract.T)*(F*norm.cdf(d1)-contract.K*norm.cdf(d2))
        return call_price
    
    def BSpricePut(self, dynamics, contract):
        F = dynamics.S*np.exp(dynamics.rGrow*contract.T)
        std = dynamics.sigma*np.sqrt(contract.T)
        d1 = np.log(F/contract.K)/std+std/2
        d2 = d1-std
        # call_price =  np.exp(-dynamics.r*contract.T)*(F*norm.cdf(d1)-contract.K*norm.cdf(d2))
        # put_price = call_price - np.exp(-dynamics.r*contract.T)*(F-contract.K)
        put_price = np.exp(-dynamics.r*contract.T)*(contract.K*norm.cdf(-d2)-F*norm.cdf(-d1))
        return put_price

    def IV(self, dynamics, contract):
        # ignores dynamics.sigma, because this function solves for sigma.

        if contract.price is None:
            raise ValueError('Contract price must be given')

        # non-arbitrage condition
        df = np.exp(-dynamics.r*contract.T)  # discount factor
        F = dynamics.S / df
        lowerbound = np.max([0,(F-contract.K)*df])
        C = contract.price
        if C<lowerbound:
            return np.nan
        if C==lowerbound: # if C=upperbound=S=F*df, then implied vol is infinity
            return 0
        if C>=F*df:
            return np.nan

        dytry = copy(dynamics)
        # We "try" values of sigma until we find sigma that generates price C

        # First find lower and upper bounds
        sigma_try = 0.2
        while self.BSpriceCall(dytry.update_sigma(sigma_try),contract)>C:
            sigma_try /= 2
        while self.BSpriceCall(dytry.update_sigma(sigma_try),contract)<C:
            sigma_try *= 2
        hi = sigma_try
        lo = hi/2
        # We have calculated "lo" and "hi" which bound the implied volatility from below and above.
        # In other words, the implied volatility is somewhere in the interval [lo,hi].
        # Then, to calculate the implied volatility within that interval,
        # for purposes of this homework, you may either (A) write your own bisection algorithm,
        # or (B) use scipy.optimize.bisect or (C) use scipy.optimize.brentq
        # You will need to provide lo and hi to those solvers.
        # There are other solvers that do not require you to bound the solution
        # from below and above (for instance, scipy.optimize.fsolve is a useful solver).
        # However, if you are able to bound the solution (of a single-variable problem),
        # then bisection or Brent will be more reliable.

        price_diff = lambda sigma: self.BSpriceCall(dytry.update_sigma(sigma), contract) - C # the function that plug into the bisection algorithm

        # (A) write your own bisection algorithm
        def my_bisection_method(f, left, right, tol=1e-5, max_iter=100000):
            if f(left) * f(right) >= 0:
                raise ValueError("f(a) and f(b) must have opposite signs")
            
            iter_count = 0
            while (right - left) / 2 > tol:
                iter_count += 1

                mid = (left + right) / 2
                if f(mid) == 0 or (right - left) / 2 < tol:
                    # print(f"Root found after {iter_count} iterations: {c}")
                    return mid
                elif f(left) * f(mid) < 0:
                    right = mid
                else:
                    left = mid

                if iter_count >= max_iter:
                    # print(f"Maximum iterations reached. Approximate root: {c}")
                    return mid
                
            # print(f"Root found after {iter_count} iterations: {(left + right) / 2}")
            return (left + right) / 2

        impliedVolatility = my_bisection_method(price_diff, lo, hi)

        # (B) use scipy.optimize.bisect
        # impliedVolatility = bisect(price_diff, lo, hi)
        
        # (C) use scipy.optimize.brentq
        # impliedVolatility = brentq(price_diff, lo, hi)
        
        return impliedVolatility


In [52]:
call1 = CallOption(100, 0.1, 5.25)
call2 = CallOption(100, 0.2, 7.25)
call3 = CallOption(100, 0.5, 9.5)

gbm_dynamics = GBMdynamics(100, 0.05, 0.05)

In [55]:
analytic_engine = AnalyticEngine()
sigma1 = analytic_engine.IV(gbm_dynamics, call1)
sigma2 = analytic_engine.IV(gbm_dynamics, call2)
sigma3 = analytic_engine.IV(gbm_dynamics, call3)

In [60]:
def calibrated_impvol(sigma_lst, time_lst):
    cal_sigma_lst = [sigma_lst[0]]
    for i in range(1, len(sigma_lst)):
        cal_sigma = (time_lst[i]*sigma_lst[i]**2 - time_lst[i-1]*sigma_lst[i-1]**2)/(time_lst[i]-time_lst[i-1])
        cal_sigma_lst.append(np.sqrt(cal_sigma))
    return cal_sigma_lst

In [63]:
sigma_lst = [sigma1, sigma2, sigma3]
time_lst = [0.1, 0.2, 0.5]
cal_sigma_lst = calibrated_impvol(sigma_lst, time_lst)
print([round(_, 4) for _ in cal_sigma_lst])

[0.3973, 0.3622, 0.2209]


- Thus, A time-varying local volatility function can be constructed as:
$$
\sigma(t) = \begin{cases} 
0.3973 & \text{if } 0 \leq t \leq 0.1 \\
0.3622 & \text{if } 0.1 < t \leq 0.2 \\
0.2209 & \text{if } 0.2 < t \leq 0.5 \\
\end{cases}
$$ 

#### (c)

In [90]:
def calc_calibrated_impvol(time_start, time_end, cal_sigma_lst, time_lst):
    def get_cumu_vol(time_target, cal_sigma_lst, time_lst):
        if time_target == 0:
            return 0
        for i in range(len(time_lst)-1):
            if time_lst[i] <= time_target and time_lst[i+1] > time_target:
                break
        cumu_vol = time_lst[0]*cal_sigma_lst[0]**2
        for j in range(1, i+1):
            cumu_vol += (time_lst[j]-time_lst[j-1])*cal_sigma_lst[j]**2
        cumu_vol += (time_target-time_lst[i])*cal_sigma_lst[i+1]**2
        return cumu_vol
    cumu_vol_start = get_cumu_vol(time_start, cal_sigma_lst, time_lst)
    cumu_vol_end = get_cumu_vol(time_end, cal_sigma_lst, time_lst)

    sigma_cal_start = np.sqrt(1/(time_end-time_start) * (cumu_vol_end-cumu_vol_start))
    return sigma_cal_start

In [93]:
sigma_cal = calc_calibrated_impvol(0, 0.4, cal_sigma_lst, time_lst)
print(f'time-0 implied vol for expiry {0.4} is {round(sigma_cal, 4)}')

call4 = CallOption(100, 0.4)
call4_price= analytic_engine.BSpriceCall(gbm_dynamics.update_sigma(sigma_cal), call4)
print(f'call price for expiry {0.4} is {round(call4_price,4)}')

time-0 implied vol for expiry 0.4 is 0.3109
call price for expiry 0.4 is 8.7842


$$ \sigma_{t,imp} = \sqrt{\frac{1}{T-t}\int_{t}^{T}\sigma(s)^2ds} $$

In [94]:
sigma_cal_interim = calc_calibrated_impvol(0.1, 0.4, cal_sigma_lst, time_lst)
print(f'time-0.1 implied vol for expiry {0.4} is {round(sigma_cal_interim,4)}')

time-0.1 implied vol for expiry 0.4 is 0.2761
